test/csuite/timestamp_abort/main.c

```diff
@@ -121,7 +121,7 @@ thread_ts_run(void *arg)
 	WT_CURSOR *cur_stable;
 	WT_SESSION *session;
 	THREAD_DATA *td;
-	uint64_t i, last_ts, oldest_ts;
+	uint64_t i, last_ts, oldest_ts, this_ts;
```

Comment:
I think this_ts needs to be `volatile` to avoid the compiler optimizing out the local variable. @keithbostic probably has an opinion about that, since we are still working towards being on the same page about the purpose of volatile.

Reply:
I thought `volatile` was for globals so that it wasn't cached/optimized between threads.  We `WT_PUBLISH` every modification to the array so that isn't the problem.  I don't know what `volatile` does to a local stack variable, if anything.  It probably isn't harmful.  But I'll let @keithbostic have the say here.

Reply:
@sueloverso, the `WT_PUBLISH` only guarantees that all writes by the publishing before the publish are visible before the published write.  It doesn't have any effect on the code generated by compilers or scheduling within CPUs by reading threads.

The concern here is that setting a local variable `this_ts = th_ts[i]` has no semantic meaning: the compiler can (and usually does) implement that as an ordinary load from memory, but (absent some constraint like `volatile` or a barrier in the reading loop), it could equally well cache a value in a register and keep reading it (never seeing a published update), or read from memory every time the source code uses a local variable.

In this case, I think the best solution is to mark the global `th_ts` array `volatile`, so that the compiler has to generate a read each time it accessed, and cannot assume that the value stays the same after it reads into a local variable.

Reply:
I think we need to talk because what you describe is what I'm trying to avoid and a description of the actual problem.  Marking the `th_ts` array volatile would likely make the original bug happen more frequently.  I want to be guaranteed that the value the reader first reads from `th_ts[i]` is the only value it sees for the duration of that loop.  If it changes on the writing CPU, I do not want the reader to see that updated value between the statements, but the next time through the walk of the array.  The sleep at the bottom should cause it to see updated values the next time around.

Reply:
Hi @sueloverso, sorry if it wasn't clear, but I'm suggesting we add `volatile` to `th_ts` in addition to your changes.  Without it, the local variable has no semantic meaning: the compiler could substitute re-reading the global value where it sees uses of the local variable.  Marking the variable (or the read) `volatile` prevents that.

Reply:
Thanks for that clarification.  I was totally misinterpreting at first.  That makes sense and I have pushed that change.  This should be ready for merge.